### PR TITLE
Fixed desyncs by placing some scenery.

### PR DESCRIPTION
**Pull request by ZehM4tt**
Also refactored the DEBUG_DESYNC logging to improve further searches in desyncs.

### CODE REVIEW

File: src/openrct2/ride/ride.c [hunk 1]
```diff
@@ -1090,9 +1090,12 @@ void ride_clear_for_construction(sint32 rideIndex)
     ride->lifecycle_flags &= ~(RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN);
     ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAIN | RIDE_INVALIDATE_RIDE_LIST;
 
+    // Fixes another MP desync where it left the rating set to the previous run causing peeps to pick different rides.
+    invalidate_test_results(rideIndex);
+
     ride_remove_cable_lift(ride);
     ride_remove_vehicles(ride);
-
+
```
Comment (Gymnasiast) [inline src/openrct2/ride/ride.c]: Trailing whitespace.

File: src/openrct2/ride/ride.c [hunk 2]
```diff
@@ -1090,9 +1090,12 @@ void ride_clear_for_construction(sint32 rideIndex)
     ride->lifecycle_flags &= ~(RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN);
     ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAIN | RIDE_INVALIDATE_RIDE_LIST;
 
+    // Fixes another MP desync where it left the rating set to the previous run causing peeps to pick different rides.
+    invalidate_test_results(rideIndex);
```
Comment (Gymnasiast) [inline src/openrct2/ride/ride.c]: Just clearing for construction should _not_ invalidate the test results. It's very annoying to the user and prevents the 'double close and open' trick. If there is absolutely no other way, at least wrap this in an if that checks for multiplayer.

I'd also like to know how this would cause desyncs. Just clearing for construction does not affect the rating, only actually changing the ride will.

File: src/openrct2/network/network.cpp [hunk 3]
```diff
@@ -1136,7 +1142,12 @@ void Network::Server_Send_TICK()
     // but debug version can check more often.
     static sint32 checksum_counter = 0;
     checksum_counter++;
+#ifdef DEBUG_DESYNC 
+    // Check every tick in debug version.
+    if (checksum_counter >= 1) {
+#else
```
Comment (duncanspumpkin) [inline src/openrct2/network/network.cpp]: This is pretty horrible can't you just have different values defined for the different compile modes

File: src/openrct2/peep/peep.c [hunk 4]
```diff
@@ -7835,6 +7946,8 @@ static void peep_stop_purchase_thought(rct_peep* peep, uint8 ride_type){
 
 void peep_set_map_tooltip(rct_peep *peep)
 {
+    //scenario_log("%s(%08X)\n", __FUNCTION__, peep->id);
```
Comment (duncanspumpkin) [inline src/openrct2/peep/peep.c]: Don't introduce commented out code unless there is a reason

File: src/openrct2/peep/peep.c [hunk 5]
```diff
@@ -11797,7 +11982,6 @@ static bool peep_should_go_on_ride(rct_peep *peep, sint32 rideIndex, sint32 entr
 
         // Assuming the queue conditions are met, peeps will always go on free transport rides.
         // Ride ratings, recent crashes and weather will all be ignored.
-        money16 ridePrice = ride_get_price(ride);
         if (!(RideData4[ride->type].flags & RIDE_TYPE_FLAG4_TRANSPORT_RIDE) || ride->value == 0xFFFF || ridePrice != 0) {
```
Comment (duncanspumpkin) [inline src/openrct2/peep/peep.c]: Why this change?

File: src/openrct2/world/footpath.c [hunk 6]
```diff
@@ -346,8 +347,19 @@ static money32 footpath_place_real(sint32 type, sint32 x, sint32 y, sint32 z, si
         return MONEY32_UNDEFINED;
     }
 
-    if (flags & GAME_COMMAND_FLAG_APPLY)
-        footpath_interrupt_peeps(x, y, z * 8);
+    if ((flags & GAME_COMMAND_FLAG_APPLY)) {
```
Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: I would modify this to make it the same for network or single player. Peeps shouldn't be effected by ghosts for the paths

File: src/openrct2/peep/peep.c (hunk 5 above)
Comment (Gymnasiast) [inline src/openrct2/peep/peep.c]: He moved it to the top of the function, in order to log it.

File: src/openrct2/ride/ride.c [hunk 7]
```diff
@@ -1090,8 +1090,12 @@ void ride_clear_for_construction(sint32 rideIndex)
     ride->lifecycle_flags &= ~(RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN);
     ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAIN | RIDE_INVALIDATE_RIDE_LIST;
 
-    // Fixes another MP desync where it left the rating set to the previous run causing peeps to pick different rides.
-    invalidate_test_results(rideIndex);
+    // Open circuit rides will go directly into building mode where it would normally clear the stats,
+    // however this causes desyncs since its directly ran from the panel and other clients would not get it.
+    // So we have it in here since this function will be called on all clients and shall only be invalidated in multiplayer.
```
Comment (Gymnasiast) [inline src/openrct2/ride/ride.c]: Minor nitpick: I'd change this to:
```
// Open circuit rides will go directly into building mode (creating ghosts) where it would normally clear the stats,
// however this causes desyncs since it's directly run from the window and other clients would not get it.
// To prevent these problem, unconditionally invalidate the test results on all clients in multiplayer games.
```

File: src/openrct2/network/network.cpp (hunk 3 above)
Comment (ZehM4tt) [inline src/openrct2/network/network.cpp]: I've changed it in https://github.com/OpenRCT2/OpenRCT2/pull/5621/commits/a5573aada98aab5a318f361eb10c3ca08a62aac6

File: src/openrct2/ride/ride.c (hunk 7 above)
Comment (ZehM4tt) [inline src/openrct2/ride/ride.c]: I'll fix it.

File: src/openrct2/ride/ride.c (hunk 1 above)
Comment (ZehM4tt) [inline src/openrct2/ride/ride.c]: Should be gone.

File: src/openrct2/world/footpath.c (hunk 6 above)
Comment (ZehM4tt) [inline src/openrct2/world/footpath.c]: I still wonder if its part of the original game to behave like that or some actual bug that shouldn't exist, for offline gameplay its not a game breaking thing.

File: src/openrct2/ride/ride.c [hunk 8]
```diff
@@ -1090,9 +1090,9 @@ void ride_clear_for_construction(sint32 rideIndex)
     ride->lifecycle_flags &= ~(RIDE_LIFECYCLE_BREAKDOWN_PENDING | RIDE_LIFECYCLE_BROKEN_DOWN);
     ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAIN | RIDE_INVALIDATE_RIDE_LIST;
 
-    // Open circuit rides will go directly into building mode where it would normally clear the stats,
-    // however this causes desyncs since its directly ran from the panel and other clients would not get it.
-    // So we have it in here since this function will be called on all clients and shall only be invalidated in multiplayer.
+    // Open circuit rides will go directly into building mode (creating ghosts) where it would normally clear the stats,
+    // however this causes desyncs since it's directly run from the window and other clients would not get it.
+    // To prevent these problem, unconditionally invalidate the test results on all clients in multiplayer games.
```
Comment (Gymnasiast) [inline src/openrct2/ride/ride.c]: Oh bugger, I made a typo. I meant to write:
```
To prevent these problems
```

Comment (ZehM4tt) [inline src/openrct2/ride/ride.c]: Here :D

File: src/openrct2/world/footpath.c (hunk 6 above)
Comment (rwjuk) [inline src/openrct2/world/footpath.c]: I think it's an original bug. If you put down a path ghost, guests will often try to walk across it - I'd say the ghosts should be ignored under all circumstances.

Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: Agreed.

Comment (ZehM4tt) [inline src/openrct2/world/footpath.c]: Then I will clean up this part, thanks for the info.

File: src/openrct2/world/footpath.c [hunk 9]
```diff
@@ -346,7 +346,7 @@ static money32 footpath_place_real(sint32 type, sint32 x, sint32 y, sint32 z, si
         return MONEY32_UNDEFINED;
     }
 
-    if (flags & GAME_COMMAND_FLAG_APPLY)
+    if ((flags & GAME_COMMAND_FLAG_APPLY) && (flags & GAME_COMMAND_FLAG_7) && !(flags & GAME_COMMAND_FLAG_GHOST))
```
Comment (janisozaur) [inline src/openrct2/world/footpath.c]: what is `GAME_COMMAND_FLAG_7` used here?

File: src/openrct2/peep/peep.c [hunk 10]
```diff
@@ -12103,7 +12103,7 @@ static void peep_pick_ride_to_go_on(rct_peep *peep)
         // Consider rides that peep hasn't been on yet
         sint32 i;
         FOR_ALL_RIDES(i, ride) {
-            if (!peep_has_ridden(peep, i)) {
+            if (!peep_has_ridden(peep, i) && ride->status == RIDE_STATUS_OPEN) {
```
Comment (Gymnasiast) [inline src/openrct2/peep/peep.c]: This seems to be unrelated.

File: src/openrct2/world/footpath.c (hunk 9 above)
Comment (Gymnasiast) [inline src/openrct2/world/footpath.c]: Why the introduction of GAME_COMMAND_FLAG_7?

File: src/openrct2/peep/peep.c (hunk 10 above)
Comment (janisozaur) [inline src/openrct2/peep/peep.c]: Indeed

File: src/openrct2/world/footpath.c (hunk 9 above)
Comment (rwjuk) [inline src/openrct2/world/footpath.c]: Also, if you know what the flag is/does, could you rename it appropriately.

Comment (duncanspumpkin) [inline src/openrct2/world/footpath.c]: `GAME_COMMAND_FLAG_7` is set only if you are placing footpath scenery this seems like a mistake. You should interrupt peeps on a footpath place as well since you might be deleteing a paths seat. I'm also going to make an issue saying name `GAME_COMMAND_FLAG_7`

File: src/openrct2/peep/peep.c (hunk 10 above)
Comment (ZehM4tt) [inline src/openrct2/peep/peep.c]: Yes sorry, this happened while trying to figure out some other desync.

File: src/openrct2/world/footpath.c (hunk 9 above)
Comment (ZehM4tt) [inline src/openrct2/world/footpath.c]: When the ghost piece was active and I placed it somewhere it started to constantly spam the command without the ghost flag, so I made it only invalidates when the flag is present. I will take a look into it.

Comment (ZehM4tt) [inline src/openrct2/world/footpath.c]: Looks like I made a mistake, it is enough to only check for the ghost flag.
